File: full_write.py
#Output is "full_data" containing voltages at time intervals.

import math
import numpy as np
import matplotlib.pyplot as plt
import time

#Constants
timestep = 10**(-2)         #How many computations per ms of data
L = 3.0                     #Length of the heart fiber in cm
N = 300                     #Number of "cells" in the fiber
spacestep = L/float(N)      #cm size per array
v_crit = .13                #Coefficient for voltage model
t_in = .1                   #Coefficient for cell gate model
t_out = 2.4                 #""
t_open = 130.0              #""
t_close = 150.0             #""
K = .001                    #Coefficient for numerical stability
BCL = 400.0                 #Mean of heart rate
sd = 20.0                   #SD of heart rate

#Resolution variable
#Determines how often to produce plot of voltage
#Recommend around 600, can be varied dependent on size of computation
resolution = 10

#Gate condition
def g(v,h):
    if v <= v_crit:
        return (1 - h)/t_open
    else:
        return (-h)/t_close

#Single cell voltage condition
def f(v,h):
    return h*(v**2)*(1-v)/t_in - (v/t_out)

#Creates a heart "pump" by charging first few cells  
def v_init(x):
    if x <= 4*spacestep:
        return .8
    elif x >= L-4*spacestep:
        return .8
    else:
        return 0

#initial condition for cell gates       
def h_init(x):
    return 1

#returns gate constant, h, at time m+1
def update_h(v,h):
    return g(v,h)*timestep + h

#returns voltage at spot k, time m_1, for middle heart cells
#must be given time m and location k
def stdupdate_v(k,V,H):
    return (((V[k+1] - 2*V[k] + V[k-1])*(K/(spacestep**2))) + f(V[k],H[k]))* timestep + V[k]

#boundary conditions    
def lupdate_v(k,V,H):
    return (((2*V[1] - 2*V[0])*(K/(spacestep**2))) + f(V[0],H[0]))* timestep + V[0]
    
def rupdate_v(k,V,H):
    return (((2*V[N-1] - 2*V[N])*(K/(spacestep**2))) + f(V[N],H[N]))* timestep + V[N]
    
#generates next time of heart pump with some natural variance
def rand():
    return int(np.random.normal(BCL,sd))


def main(T_output):   
    print "Initializing..."
    
    L_var = rand()                      #determines next pacemaker pulse time
    R_var = rand()

    T = int(T_output/timestep)          #number of ms to compute divided by granularity

    frames = math.floor(T/resolution)   #number of plots to be created

    #Conditions must satisfy this for numerical stability
    if (K*timestep)/(spacestep**2) > .5:
        print "CFL condition not met"
        return 0
    
    #data arrays for voltage
    width = N+1
    
    V_old = [0 for x in range(width)]
    V_new = [0 for x in range(width)]
    
    #data arrays for gate function
    H_old = [0 for x in range(width)]
    H_new = [0 for x in range(width)]
    
    #fill in the first array with the initial voltage
    #do once each for N cells
    for i in range(0,N+1):
        x = i*spacestep
        V_old[i] = v_init(x)
    
    V_new = V_old
    
    ary1 = []
    for i in range(0,N+1):
        ary1.append(i*spacestep)
    
    #Initialize H values
    for i in range(0,N+1):
        H_old[i] = 1
    
    print "Loading frame:"

    m = 0
    
    while m < T+1:
        #occasionally append voltage data in txt file
        if m % resolution == 0:
            # plt.plot(ary1,V_old)
            f = open("full_data","a")
            for i, array in enumerate(V_old):
                s = str(V_old[i]) + ' '
                f.write(s)
            f.write('\n')
            n = int(m/resolution)
            # plt.savefig('data_%d.png'%(n,))
            # plt.gcf().clear()

            print n,"/",int(frames)
            
        #fill in interior grid points
        for k in range(1,N):
            V_new[k] = stdupdate_v(k,V_old,H_old)
            
        #fill in left boundary
        V_new[0] = lupdate_v(k,V_old,H_old)
        #fill in right boundary
        V_new[N] = rupdate_v(k,V_old,H_old)
        
        #update door constants
        for k in range(0,N+1):    
            H_new[k] = update_h(V_old[k],H_old[k])
         
        #if time for a new pump, add voltage   
        if int((m+1)*timestep) == L_var:
            for i in range(0,5):
                V_new[i] = .8
            L_var = rand()+L_var
        
        if int((m+1)*timestep) == R_var:
            for i in range(N-4,N+1):
                V_new[i] = .8
            R_var = rand()+R_var

        V_old = V_new
        H_old = H_new 
        
        m += 1

start = time.time()
#Value in main is number of milliseconds to compute
main(10)        #using 60000 gives 1 minute of simulation
print "Time: ", time.time()-start

File: C_version/full_write.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

// global variables
float T_output = 60000;
int resolution = 600;
float timestep = 0.01;
float L = 3.0;
int N = 300;
float spacestep = 0.01; // adjust manually!!!
float v_crit = .13;
float t_in = .1;
float t_out = 2.4;
float t_open = 130.0;
float t_close = 150.0;
float K = .001;
float BCL = 400.0;
float sd = 20.0;

float g(float v, float h) {
	if (v <= v_crit) {
		return((1-h)/t_open);
	}
	else {
		return(-1*h/t_close);
	}
}

// single cell voltage condition
float f(float v, float h) {
	return(h*v*v*(1-v)/t_in - v/t_out);
}

// initial values
float v_init(float x) {
	if (x <= 4*spacestep) {
		return(0.8);
	}
	else if (x >= L-4*spacestep) {
		return(0.8);
	}
	else {
		return(0);
	}
}

float h_init(float x) {
	return(1);
}

// returns h^(m+1)
float update_h(float v, float h) {
	return(g(v,h)*timestep + h);
}

// returns v^(m+1) at spot k on middle parts given time m and spot k
float stdupdate_v(int k, float *V, float *H) {
	return((((V[k+1] - 2*V[k] + V[k-1])*(K/(spacestep*spacestep))) + f(V[k],H[k]))* timestep + V[k]);
}
    
float lupdate_v(float *V, float *H) {
    return((((2*V[1] - 2*V[0])*(K/(spacestep*spacestep))) + f(V[0],H[0]))* timestep + V[0]);
}

float rupdate_v(float *V,float *H) {
    return((((2*V[N-1] - 2*V[N])*(K/(spacestep*spacestep))) + f(V[N],H[N]))* timestep + V[N]);
}
    
// generate random fluctuations (using Central Limit Thm)
int rand_fluc(float BCL, float sd) {
	float x = 0;
	for (int i = 0; i < 25; i++) {
		x += rand();
	}
	x -= 25/2.0;
	x /= sqrt(25/12.0);
    return((int)(sd*x + BCL));
}

// copy one float array to another float array
void copy_arr(float *A, float *B, int n) {
	for (int i = 0; i < n; i ++) {
		A[i] = B[i];
	}
}

int main(int argc, char **argv) {

	printf("Initializing...");

	float L_var = rand_fluc(BCL,sd);
	float R_var = rand_fluc(BCL,sd);

	int T = (int) T_output/timestep;
	int frames = T/resolution;

	if (K*timestep/(spacestep*spacestep) > 0.5) {
		printf("CFL condition not met");
		return(0);
	}

	int m = 0;

	// stores voltage, v, at each location, x, in an array
	// (stores one array per timestamp)
	int width = N+1;
	int height = T_output + 2;

	float V_old[width];
	float V_new[width];

	// array for gate function
	float H_old[width];
	float H_new[width];

	// fill in the 1st array w/ initial voltage
	for (int i = 0; i < width; i++) {
		V_old[i] = v_init(i*spacestep);
	}

	float ary1[width];
	for (int i = 0;i < N; i++) {
		ary1[i] = i*spacestep;
	}

	// initialize H values
	for (int i = 0;i < N; i++) {
		H_old[i] = 1;
	}

	// prepare file for writing
	FILE *f = fopen("fulldatac", "wb");

	printf("Loading frame:");

  	// start timer
  	clock_t begin = clock();

	while (m < T+1) {

		// write to file (frequency depends on resolution)
		if (m % resolution == 0) {
			for (int i = 0; i < width-1; i++) {
			fprintf(f,"%f\t",V_old[i]);
			}
			fprintf(f,"%f\n",V_old[width-1]);
		}	


		// fill in interior grid points
		for (int i = 1; i < N; i++) {
			V_new[i] = stdupdate_v(i,V_old,H_old);
		}

		// fill in left boundary
		V_new[0] = lupdate_v(V_old,H_old);
		// fill in right boundary
		V_new[N] = rupdate_v(V_old,H_old);

		for (int i = 0; i < width; i++) {
			H_new[i] = update_h(V_old[i],H_old[i]);
		}

		if ((int)((m+1)*timestep) == R_var) {
			for (int i = N-4; i < width; i++) {
				V_new[i] = 0.8;
			}
			R_var = rand_fluc(BCL,sd) + R_var;
		}

		// update V, H arrays
		copy_arr(V_old,V_new,width);
		copy_arr(H_old,H_new,width);

		m += 1;
	}

	// end timer
	clock_t end = clock();
	double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
	printf("%f",time_spent);

	fclose(f);
	
	return(0);
}
